cpp/src/arrow/array/concatenate.cc

```diff
@@ -163,6 +163,46 @@ static Status PutOffsets(const std::shared_ptr<Buffer>& src, Offset first_offset
   return Status::OK();
 }
 
+struct DictionaryConcatenate {
```

Comment:
The rest of the concatenation functions simply memcpy'd the buffers.  However, the dictionary concatenation needs to map buffers to potentially new index values.  As a result, this function needs to know the type of the buffer for the reinterpret case on line 190.  Also, the fact that memo table indices are 32 bit and dictionary indices could be 64 bit is a potential problem but one that already existed and it seems unlikely that a dictionary array would be used when there are 4B unique values.

Reply:
Done